EIPS/eip-draft_request_id.md

```diff
@@ -0,0 +1,135 @@
+
+The `request_id` is a simple integer set by the client when it makes the request. On the responding side, the exact same `request_id` from the incoming request is put back into the response object.
+
+This change allows the requesting client to match incoming responses **directly** back to their pending requests without going through all of the pending requests to check if they might match based on the response data.
+
+The selected request/response pair serves as an example for many similar request/response pairs in the `eth` networking protocol.
+
+## Motivation
+<!--The motivation is critical for EIPs that want to change the Ethereum protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the EIP solves. EIP submissions without sufficient motivation may be rejected outright.-->
+The lack of request identifiers in the request / response paris of the `eth` protocol puts unnecessary burden of code complexity into every Ethereum client. It also makes the communication slightly less efficient. Another argument can be made that the addition of request identifiers makes the protocol more aligned with the `les` protocol which **does** already defines request identifiers for each request / response pair.
+
+## Specification
+<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (go-ethereum, parity, cpp-ethereum, ethereumj, ethereumjs, and [others](https://github.com/ethereum/wiki/wiki/Clients)).-->
+
+Change the following message types in the `eth` protocol:
+
+* `GetBlockHeaders (0x03)`
+   * **Current:** `[block: {P, B_32}, maxHeaders: P, skip: P, reverse: P in {0, 1}]`
+   * **Then**: `[request_id: P, [block: {P, B_32}, maxHeaders: P, skip: P, reverse: P in {0, 1}]]`
+* `BlockHeaders (0x04)`
+   * **Current:** `[blockHeader_0, blockHeader_1, ...]`
+   * **Then**: `[request_id: P, [blockHeader_0, blockHeader_1, ...]]`
+* `GetBlockBodies (0x05)`
+   * **Current:** `[hash_0: B_32, hash_1: B_32, ...]`
+   * **Then**: `[request_id: P, [hash_0: B_32, hash_1: B_32, ...]]`
+* `BlockBodies (0x06)`
+   * **Current:** `[hash_0: B_32, hash_1: B_32, ...]`
+   * **Then**: `[request_id: P, [hash_0: B_32, hash_1: B_32, ...]]`
+* `GetNodeData (0x0d)`
+   * **Current:** `[hash_0: B_32, hash_1: B_32, ...]`
+   * **Then**: `[request_id: P, [hash_0: B_32, hash_1: B_32, ...]]`
+* `NodeData (0x0e)`
+   * **Current:** `[value_0: B, value_1: B, ...]`
+   * **Then**: `[request_id: P, [value_0: B, value_1: B, ...]]`
+* `GetReceipts (0x0f)`
+   * **Current:** `[blockHash_0: B_32, blockHash_1: B_32, ...]`
+   * **Then**: `[request_id: P, [blockHash_0: B_32, blockHash_1: B_32, ...]]`
+* `Receipts (0x10)`
+   * **Current:** `[[receipt_0, receipt_1], ...]`
+   * **Then**: `[request_id: P, [[receipt_0, receipt_1], ...]]`
+
```

Comment:
I think it's important to mention the range of valid request IDs (whether it's a 64bit int or a 256bit one). I'm not sure what `P` means any more, but I'm assuming 64.

Also I think it would be a very good thing to support *random* IDs (i.e. explicitly state that these IDs don't have to be sequential). The reason is that if we always start the requests from 0, then a peer that keeps disconnecting and reconnecting will get issues the same IDs and will reply with the same ID, so calling code will have a harder time mixing and matchins responses across disconnects (geth has a whole lot of logic around this). With random IDs, the probability of 2 in-fight (or even recent) requests clashing is effectively nil.

I'd also say that it's allowed to have duplicate IDs (i.e. same in-flight request with the same ID). The reason is that requesters don't have to care about the ID if they don't want to. If they randomize and accidentally duplicate it's not an issue. And requesting diferent things (header, body, state) with the same ID would be valid (since there's no way you'd mix them up anyway).

Reply:
These are fine remarks. What you describe is also what I had in **mind** but I agree that the spec should be clear about that :+1: I'll get that updated.